client/components/QueryListDropdown/QueryListDropdown.tsx

```diff
@@ -17,6 +17,14 @@ type Props<Item> = {
 	itemRenderer: ItemRenderer<Item>;
 	usePortal?: boolean;
 	query?: string;
+	newItemFromQuery?: (query: string) => Item;
```

Comment:
Since this is getting wordy, one trick I'd suggest using is to pull these definitions off the `QueryList` component:

```ts
type SharedQueryListProps = Pick<
	React.ComponentProps<typeof QueryList>,
	'createNewItemFromQuery' | 'createNewItemRenderer'
>;

type Props<Item> = SharedQueryListProps & { ... };
``` 

Breaking this down:

- The [TypeScript `typeof` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof) gives you the type that TS has inferred for a variable. This is, sadly, not to be confused with the [runtime `typeof` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof) which does something totally different. But the former can only appear in type expressions, and the latter only in runtime code.
- `React.ComponentProps<T>` tells you the props of a React component with type `T`.
- `Pick<T>` you've already used here!

This produces the original prop names from `QueryList` with the "create" prefix, which I think we should keep anyway. There are probably even more props in this definition that could be picked from the `QueryList` definition instead of manually written out!